libmpv/ios/ViewController.m

```diff
@@ -0,0 +1,211 @@
+@import GLKit;
+@import OpenGLES;
+
+#import "ViewController.h"
+
+
+#import <mpv/client.h>
+#import <mpv/opengl_cb.h>
+
+#import <stdio.h>
+#import <stdlib.h>
+
+
+static inline void check_error(int status)
+{
+    if (status < 0) {
+        printf("mpv API error: %s\n", mpv_error_string(status));
+        exit(1);
+    }
+}
+
+static void *get_proc_address(void *ctx, const char *name)
+{
+    CFStringRef symbolName = CFStringCreateWithCString(kCFAllocatorDefault, name, kCFStringEncodingASCII);
+    void *addr = CFBundleGetFunctionPointerForName(CFBundleGetBundleWithIdentifier(CFSTR("com.apple.opengles")), symbolName);
+    CFRelease(symbolName);
+    return addr;
+}
+
+static void glupdate(void *ctx);
+
+@interface MpvClientOGLView : GLKView
+    @property mpv_opengl_cb_context *mpvGL;
+@end
+
+@implementation MpvClientOGLView {
+    GLint defaultFBO;
+}
+    
+- (void)awakeFromNib
+{
+    [super awakeFromNib];
+    self.context = [[EAGLContext alloc] initWithAPI:kEAGLRenderingAPIOpenGLES3];
+    [EAGLContext setCurrentContext:self.context];
+
+    // Configure renderbuffers created by the view
+    self.drawableColorFormat = GLKViewDrawableColorFormatRGBA8888;
+    self.drawableDepthFormat = GLKViewDrawableDepthFormatNone;
+    self.drawableStencilFormat = GLKViewDrawableStencilFormatNone;
+    
+    defaultFBO = -1;
+}
+    
+- (void)fillBlack
+{
+    glClearColor(0, 0, 0, 0);
+    glClear(GL_COLOR_BUFFER_BIT);
+}
+    
+- (void)drawRect
+{
+    if (defaultFBO == -1)
+    {
+        glGetIntegerv(GL_FRAMEBUFFER_BINDING, &defaultFBO);
+    }
+
+    if (self.mpvGL)
+    {
+        mpv_opengl_cb_draw(self.mpvGL,
+                           defaultFBO,
+                           self.bounds.size.width * self.contentScaleFactor,
+                           -self.bounds.size.height * self.contentScaleFactor);
+    }
+    else
+    {
+        glFlush();
+    }
+}
+
+- (void)drawRect:(CGRect)rect
+{
+    [self drawRect];
+}
+
+@end
+
+
+
+static void wakeup(void *);
+
+
+static void glupdate(void *ctx)
+{
+    MpvClientOGLView *glView = (__bridge MpvClientOGLView *)ctx;
+    // I'm still not sure what the best way to handle this is, but this
+    // works.
```

Comment:
That should be correct. Basically it wants you to wake up your rendering thread when this callback is called. Using the dispatch queue is a decent way to achieve this. In the worst case this could potentially lead to draw calls being queued up, but this probably won't happen.